master/pkg/searcher/asha.go

```diff
@@ -0,0 +1,244 @@
+	// the bracket of successive halving is finished.
+	rungIndex := s.trialRungs[requestID]
+	rung := s.rungs[rungIndex]
+	rung.outstandingTrials--
+	addedNewTrial := false
+
+	ops := make([]Operation, 0)
+	// If the trial has completed the top rung's validation, close the trial.
+	if rungIndex == s.AsyncHalvingConfig.NumRungs-1 {
+		s.trialsCompleted++
+		if !s.earlyExitTrials[requestID] {
+			ops = append(ops, NewClose(requestID))
+		}
+	} else {
+		// This is not the top rung, so do promotions to the next rung.
+		nextRung := s.rungs[rungIndex+1]
+		for _, promotionID := range rung.promotionsAsync(
+			requestID,
+			metric,
+			s.AsyncHalvingConfig.Divisor,
+		) {
+			s.trialRungs[promotionID] = rungIndex + 1
+			nextRung.outstandingTrials++
+			if !s.earlyExitTrials[promotionID] {
+				ops = append(
+					ops, trainAndValidate(promotionID, rung.stepsNeeded, nextRung.stepsNeeded)...)
+				addedNewTrial = true
+			} else {
+				step := s.rungs[rungIndex+1].stepsNeeded
+				wkld := Workload{
+					Kind:         ComputeValidationMetrics,
+					ExperimentID: message.ExperimentID,
+					TrialID:      message.TrialID,
+					StepID:       step,
+				}
+
+				// We make a recursive call that will behave the same
+				// as if we'd actually run the promoted job and received
+				// the worse possible result in return.
+				return s.promoteAsync(ctx, promotionID, wkld, ashaExitedMetricValue)
+			}
+		}
+	}
+
+	allTrials := len(s.trialRungs)
+	if !addedNewTrial && allTrials < s.maxTrials {
+		create := NewCreate(
+			ctx.rand, sampleAll(ctx.hparams, ctx.rand), model.TrialWorkloadSequencerType)
+		ops = append(ops, create)
+		ops = append(ops, trainAndValidate(create.RequestID, 0, s.rungs[0].stepsNeeded)...)
+		s.rungs[0].outstandingTrials++
+		s.trialRungs[create.RequestID] = 0
+	}
+
+	// Only close out trials once we have reached the maxTrials for the searcher.
+	if allTrials == s.maxTrials {
+		ops = append(ops, s.closeOutRungs()...)
+	}
+	return ops
+}
+
+// closeOutRungs closes all remaining unpromoted trials in any rungs that have no more outstanding
+// trials.
+func (s *asyncHalvingSearch) closeOutRungs() []Operation {
+	ops := make([]Operation, 0)
+	for _, rung := range s.rungs {
+		if rung.outstandingTrials > 0 {
+			break
+		}
+		for tid, trialMetric := range rung.metrics {
+			if !trialMetric.promoted && !trialMetric.closed {
+				s.trialsCompleted++
+				if !s.earlyExitTrials[trialMetric.requestID] {
+					ops = append(ops, NewClose(trialMetric.requestID))
+				}
+				rung.metrics[tid].closed = true
+				ops = append(ops, NewClose(trialMetric.requestID))
+			}
+		}
+	}
+	return ops
+}
+
+func (s *asyncHalvingSearch) progress(workloadsCompleted int) float64 {
+	allTrials := len(s.trialRungs)
+	// Give ourselves an overhead of 20% of maxTrials when calculating progress.
+	progress := float64(allTrials-s.rungs[0].outstandingTrials) / (1.2 * float64(s.maxTrials))
```

Comment:
question: is this what we want? It only uses the first rung to calculate progress? What if there are no outstanding trials in the 0th run but there are in other ones?

Reply:
when len(rung[0]) < maxTrials, we look at how many trials have been completed on the bottom rung, which is len(s.trialRungs) - s.rungs[0].outstandingTrials and divide by 1.2 * s.maxTrials.  when we've submitted maxTrials to the bottom rung, we switch over to how many trials have been closed.